#!/usr/bin/env bash
# =================================================================================
#                                   CONFIG                                        #
#                Please read this config file carefully                           #
# =================================================================================


###################################################################################
#                                Required modules
###################################################################################

#CONDA_INIT="eval \"\$(conda shell.bash hook)\""
CONDA_INIT=':'
CONDA_ENV_SGA="/cfs/klemming/projects/supr/naiss2025-23-301/tools/miniconda3/envs/SGA"             
FASTP_MODULE="fastp/0.24.0"
PRINSEQ_LITE="/cfs/klemming/projects/snic/sediment_paleogenomics/tools/prinseq-lite-0.20.4/prinseq-lite.pl"
KRAKEN2_MODULE="kraken2/2.1.2"
BOWTIE2_MODULE="bowtie2/2.5.4"
SAMTOOLS_MODULE="samtools/1.20"
CONDA_ENV_FILTERBAM="/cfs/klemming/projects/supr/naiss2025-23-301/tools/miniconda3/envs/bam-filter"
CONDA_ENV_ngsLCA="/cfs/klemming/projects/supr/naiss2025-23-301/tools/miniconda3/envs/ngsLCA"
PDC_MODULE="PDC"                 # only needed for Kraken2 section using $PDC_TMP
KRONATOOLS_MODULE="kronatools/2.8.1"

# ---- bamdam from Python venv ----
# Optional: module to load before activating the venv (kept generic).
BAMDAM_PYTHON_MODULE="python"
BAMDAM_VENV="/cfs/klemming/projects/supr/sllstore2017093/sediment/nathan/bamdam-env"
#Install bamdam
#python -m venv bamdam-env
#source bamdam-env/bin/activate
#pip install bamdam

###################################################################################
#                        Input files
###################################################################################

# 1) If you want to run the whole pipeline from raw sequencing data
# Point the path of your raw sequencing files or to a list containing the path of each samples to be processed
#If a list is provided, it will perform the samples contained in it preferentially and READS_GLOB won't be perform
READS_GLOB="/cfs/klemming/projects/supr/sllstore2017093/sediment/nathan/samples/*.fastq.gz" #e.g. READS_GLOB="/data/project/*_1.fastq.gz" or READS_GLOB="/data/project/*_R1.fastq.gz"
READS_LIST="/cfs/klemming/projects/supr/sllstore2017093/sediment/nathan/List_fastq.txt"   # e.g. "/data/project/my_fastqs.list"

KRAKEN_REQUIRE_PRIMARY=1  #If you want to run Kraken GTDB on every sga or prinseq files present in the respective folder, set it to 0
MAP_REQUIRE_PRIMARY=1 #If you want to run mapping on every kraken files present in the kraken folder, set it to 0

# 2) If you want to run from any other steps, with files not generated by the pipeline, provide a list of files from the step before with absolute path 
# Note: In this case, only the samples provided in the external files list will be performed, not the one provided from sequencing data to the pipeline

# Format: either one absolute path per line, or "SAMPLE<TAB>/abs/path/to/SAMPLE_merged.fastq.gz"
OVERRIDE_LIST_FASTP="" # Give absolute path - In the case you are not using fastp to preprocess the reads
OVERRIDE_LIST_SGA="" #"/cfs/klemming/projects/supr/sllstore2017093/sediment/nathan/pipeline_metage_out/01_fastp/List_SGA.txt" # Give absolute path
OVERRIDE_LIST_PRINSEQ="" #"/cfs/klemming/projects/supr/sllstore2017093/sediment/nathan/pipeline_metage_out/01_fastp/List_Prinseq.txt"
OVERRIDE_LIST_KRAKEN="" #"/cfs/klemming/projects/supr/sllstore2017093/sediment/nathan/pipeline_metage_out/02_sga/List_Kraken.txt" # Give absolute path
OVERRIDE_LIST_MAPPING="" #"/cfs/klemming/projects/supr/sllstore2017093/sediment/nathan/pipeline_metage_out/03_kraken_gtdb/List_mapping.txt" # Give absolute path
OVERRIDE_LIST_FILTER="" #"/cfs/klemming/projects/supr/sllstore2017093/sediment/nathan/pipeline_metage_out/04_mapping/List_filter.txt" # Give absolute path

# If you already have SGA output (name: <sample>_merged.dust.rmdup.fastq.gz)
SGA_OUT_DIR=""
# If you already have Kraken2 unclassified (name: <sample>_GTDB_unclas.fastq.gz)
KRAKEN_UNCLAS_DIR=""
# If you already have mapping BAMs (name: <sample>.b2.k1000.all.sorted.bam)
MAPPING_BAM_DIR=""


###################################################################################
#                        Path to scripts and databases
###################################################################################

SCRIPTS_DIR="/cfs/klemming/projects/supr/sllstore2017093/sediment/nathan/scripts"

# NOTE: Please verify your DB versions are current; if DBs are updated, paths likely change.
#  - PhyloNorway masked bowtie2 index (prefix ending with "_"):
PHYLONORWAY="/cfs/klemming/projects/supr/sediment_paleogenomics/databases/metagenomic/PhyloNorway_masked/PhyloNorway_index/PhyloNorway_index_"
PHYLONORWAY_HEADER="/cfs/klemming/projects/supr/sediment_paleogenomics/databases/metagenomic/PhyloNorway_masked/PhyloNorway_index/PhyloNorway_sq_header.tsv"
PLASTID="/cfs/klemming/projects/supr/sediment_paleogenomics/tom/CHLOROPLAST/plastids.genomic"
MITO="/cfs/klemming/projects/snic/sediment_paleogenomics/databases/bowtie2/refseq_mito/mitochondrion.1.1.genomic"
MAM_BIRD_FISH="/cfs/klemming/projects/supr/archaeogenetics/ernjoh/mam-bird-fish_v2/mam-bird-fish_v2"

# Kraken2 GTDB 
GTDB_SRC="/cfs/klemming/pdc/software/dardel/sw-uppmax/data/Kraken2_data/prebuilt/k2_gtdb_genome_reps_20250609"

# NCBI taxonomy files
NAMES="/cfs/klemming/projects/supr/naiss2025-23-301/private/NCBI_taxonomy/names_NM.dmp"
NODES="/cfs/klemming/projects/supr/naiss2025-23-301/private/NCBI_taxonomy/nodes_NM.dmp"
ACC2TAX="/cfs/klemming/projects/supr/naiss2025-23-301/private/NCBI_taxonomy/acc2taxid_NM_PhyloNorway_MBF_3col_noNA.txt.gz"

#### Path to roots
OUT_ROOT="/cfs/klemming/projects/supr/sllstore2017093/sediment/nathan/pipeline_metage_out"
TMP_ROOT="/cfs/klemming/projects/supr/sllstore2017093/sediment/nathan/pipeline_metage_out/tmp"
LOG_ROOT="/cfs/klemming/projects/supr/sllstore2017093/sediment/nathan/pipeline_metage_out/log"

# Fastp outputs by sample will go to:     ${OUT_ROOT}/01_fastp/<sample>/
# Prinseq outputs by sample will go to:   ${OUT_ROOT}/02_prinsea/<sample>/
# SGA outputs by sample will go to:       ${OUT_ROOT}/02_sga/<sample>/
# Kraken2 outputs:                        ${OUT_ROOT}/03_kraken_gtdb/
# Mapping outputs:                        ${OUT_ROOT}/04_mapping/<runname>/
# Filtering outputs:                      ${OUT_ROOT}/05_filtering/<runname>/
# Metrics & Excel:                        ${OUT_ROOT}/99_metrics/


###################################################################################
#                              Tools activation                                   #
###################################################################################

#### 1=enable, 0=disable
ENABLE_PREPROCESS=1     # If set to 0, neither fastp or SGA will run
ENABLE_FASTP=1
ENABLE_SGA=1
ENABLE_PRINSEQ=0       # set 1 to use PRINSEQ instead of SGA         
ENABLE_KRAKEN_GTDB=1
ENABLE_MAPPING=1
ENABLE_FILTERING=1      # If set to 0, neither filterBAM or ngsLCA or bamdam will run
ENABLE_FILTERBAM=0
ENABLE_NGSLCA=1
ENABLE_BAMDAM=1
ENABLE_MMSEQS2=0
ENABLE_METRICS=0

#### Force re-run switches (0/1). When 1, submit the step even if its outputs exist.
FORCE_FASTP=1
FORCE_SGA=1
FORCE_KRAKEN=1
FORCE_MAPPING=1
FORCE_FILTER=1
 
###################################################################################
#                       Tools' parameters to precise
###################################################################################

# fastp
FASTP_OVERLAP_LEN_REQUIRE=20   # --overlap_len_require
FASTP_MIN_LENGTH=30            # -l
# SGA
SGA_DUST_THRESHOLD=4           # --dust-threshold
# PRINSEQ params (defaults mirror your snippet)
PRINSEQ_COMPLEXITY_METHOD="dust"
PRINSEQ_COMPLEXITY_THRESHOLD=4
PRINSEQ_MIN_LEN=35
PRINSEQ_DEREP="14"                      # exact fwd+rev duplicates
# Mapping run name (used to bundle outputs)
#MAP_RUN="default_run"
# bamdam parameters (used only if ENABLE_BAMDAM=1)
BAMDAM_STRANDED="ds"          # "ds" or "ss"
BAMDAM_MINREADS=5             # for bamdam krona --minreads
BAMDAM_MAXDAMAGE=0.5          # for bamdam krona --maxdamage
TOP_GENUS=10                   # Top X genus to plot for damage

###################################################################################
#                               SBATCH parameters
###################################################################################

##### Global SBATCH defaults
SBATCH_DEFAULT_ACCOUNT="naiss2025-5-78"
SBATCH_DEFAULT_PARTITION="shared"
SBATCH_DEFAULT_CPUS="32"
SBATCH_DEFAULT_MEM="28G"
SBATCH_DEFAULT_TIME="2:00:00"
SBATCH_DEFAULT_QOS=""
SBATCH_DEFAULT_EXTRA=""

##### Per-step SBATCH overrides (examples shown)
# FASTP
FASTP_SBATCH_ACCOUNT="naiss2025-5-78"
FASTP_SBATCH_PARTITION="shared"
FASTP_SBATCH_CPUS="32"
FASTP_SBATCH_MEM="28G"
FASTP_SBATCH_TIME="2:00:00"
FASTP_SBATCH_QOS=""
FASTP_SBATCH_EXTRA=""
FASTP_SBATCH_JOB_NAME="fastp_merge"

# SGA
SGA_SBATCH_ACCOUNT="naiss2025-5-78"
SGA_SBATCH_PARTITION="shared"
SGA_SBATCH_CPUS="32"
SGA_SBATCH_MEM="28G"
SGA_SBATCH_TIME="4:00:00"
SGA_SBATCH_QOS=""
SGA_SBATCH_EXTRA=""
SGA_SBATCH_JOB_NAME="SGA"

# SLURM for PRINSEQ array
PRINSEQ_SBATCH_ACCOUNT="naiss2025-5-78"
PRINSEQ_SBATCH_PARTITION="shared"
PRINSEQ_SBATCH_CPUS="32"
PRINSEQ_SBATCH_MEM="28G"
PRINSEQ_SBATCH_TIME="4:00:00"
PRINSEQ_SBATCH_QOS=""
PRINSEQ_SBATCH_EXTRA=""
PRINSEQ_SBATCH_JOB_NAME="prinseq"

# Kraken2
KRAKEN_SBATCH_ACCOUNT="naiss2025-5-78"
KRAKEN_SBATCH_PARTITION="memory" #"memory" 
KRAKEN_SBATCH_CPUS="256" #"16" 
KRAKEN_SBATCH_MEM="650G" #650G"         
KRAKEN_SBATCH_TIME="0-10:00:00" #"1-20:00:00"
KRAKEN_SBATCH_QOS=""
KRAKEN_SBATCH_EXTRA=""
KRAKEN_SBATCH_JOB_NAME="GTDB"

# Mapping
MAP_SBATCH_ACCOUNT="naiss2025-5-616" #"naiss2025-5-78"
MAP_SBATCH_PARTITION="main"   #main
MAP_SBATCH_CPUS="256" #16
MAP_SBATCH_MEM="550G"   #550G
MAP_SBATCH_TIME="0-10:00:00" #"4-00:00:00"
MAP_SBATCH_QOS=""
MAP_SBATCH_EXTRA=""
MAP_SBATCH_JOB_NAME="bowtie2"

# Filtering / ngsLCA
FILTER_SBATCH_ACCOUNT="naiss2025-5-78"
FILTER_SBATCH_PARTITION="shared"   #memory
FILTER_SBATCH_CPUS="32"
FILTER_SBATCH_MEM="28G"    #"880G"
FILTER_SBATCH_TIME="0-01:00:00"
FILTER_SBATCH_QOS=""
FILTER_SBATCH_EXTRA=""
FILTER_SBATCH_JOB_NAME="filtering_ngslca"

# Metrics
METRICS_SBATCH_ACCOUNT="naiss2025-5-78"
METRICS_SBATCH_PARTITION="shared"
METRICS_SBATCH_CPUS="1"
METRICS_SBATCH_MEM="5G"
METRICS_SBATCH_TIME="00:30:00"
METRICS_SBATCH_QOS=""
METRICS_SBATCH_EXTRA=""
METRICS_SBATCH_JOB_NAME="metrics"

#------------------------------------------------------------------------------------------------------------#
###                       IMPORTANT TO DO BEFORE TO RUN ANY MAPPING SCRIPT                                 ###
#------------------------------------------------------------------------------------------------------------#
#                 Please index all your database with bowtie2 otherwise it won't run                         #
#             Don't run more than 8-10 samples in the same time as Kraken2 might get killed                  #
#      Check carefully the disk space, so you don't have any jobs being killed for out of disk space         #
# Carefull about the input fastq.gz -> Better to add it as Sample_name_1.fastq.gz and Sample_name_2.fastq.gz #
#------------------------------------------------------------------------------------------------------------#
